### DISCUSSION TITLE
How to structure multiple layers of configuration?

### DISCUSSION THREAD

**Discussion by edward3h** [apple/pkl#332] [Q&A] [answered]
My company uses configuration from properties files, for Java applications/services, that goes something like this:

Read default values from `common.properties`.

Get a system property "profile". It contains a main profile name, followed by optional modifier names.

Read values from exactly one main profile file, specified by the name above. For example, `profiles/dev.properties` or `profiles/prod.properties`.

Read values from any modifier files specified. For example, `modifiers/canary.properties` or `modifiers/ci.properties`.

Finally, read values from `local.properties`, if it exists. This is for local developer settings and is ignored in version control.

The files read later overwrite properties defined by earlier files.

How could I structure something like this in Pkl? I currently have a `defaults.pkl`, with `dev.pkl` and `prod.pkl` using `amends "defaults.pkl"`, but I don't see how I could combine in the other layers of overriding properties.

**Comment by bioball** [comment, marked as answer]
One way to do this is to structure everything in as a mixin. A mixin is a way to have a multiple inheritance model in Pkl.

Here's a gist to demonstrate this: https://gist.github.com/bioball/8fcc848cffdf11fcbce68e420647aa1e. This example uses external properties to control what profiles and modifiers to include.

Use `pkl eval AppConfig.pkl` to get rendered results. Use external properties to determine what to include. For example:

```
pkl eval AppConfig.pkl -p modifier=ci -p profile=prod
```

This example will also import `../local/config.pkl`, but only if that file exists.

On the Java side, you can similarly set external properties when building your evaluator.

```java
public AppConfig loadAppConfig() {
  try (var evaluator = ConfigEvaluatorBuilder.preconfigured().addExternalProperty("profile", "dev").build()) {
    return evaluator.evaluate(ModuleSource.modulePath("/path/to/AppConfig.pkl")).as(AppConfig.class);
  }
}
```

Note: this is just one of the many valid options for managing complexity in configuration. Another option is via the single-inheritance "app/env/cluster" model. Under this model, you would define different folder nesting levels, where each nesting level represents one type of concern. For a demonstration of the single-inheritance model, take a look at our [advanced kubernetes example](https://github.com/apple/pkl-k8s-examples/tree/main/pkl/advanced).

**Reply by edward3h** [reply]
Thanks for the suggestions!
